### PR TITLE
[DO NOT MERGE] Code Review #1

### PR DESCRIPTION
**Pull request by duckworthd**
This is duckibot's first code review of Appacus. All code files have been removed to facilitate in-line code comments. DO! NOT! MERGE!

### CODE REVIEW

File: sampleGame/GameScene.m [hunk 1]
```diff
@@ -1,354 +0,0 @@
...
-
-#import "Abacus.h"
-#import "GameScene.h"
-#import "BeadData.h"
-
-@implementation GameScene{
-    NSMutableArray *_columnBeads; //stores all beads as SKSpriteNode objs
-    
-    //stores currentColumnValues
-    int currentValue;
```
Comment (duckworthd) [inline sampleGame/GameScene.m]: Stick to one convention -- underscores or no underscores. Readers will be able to distinguish in-function local variables from instance variables easily.

File: sampleGame/GameScene.m [hunk 2]
```diff
@@ -1,354 +0,0 @@
...
-//
-
-#import "Abacus.h"
-#import "GameScene.h"
-#import "BeadData.h"
-
-@implementation GameScene{
-    NSMutableArray *_columnBeads; //stores all beads as SKSpriteNode objs
-    
-    //stores currentColumnValues
```
Comment (duckworthd) [inline sampleGame/GameScene.m]: I like that this variable has a comment describing what it's used for :). Please do the same for others?

File: sampleGame/GameScene.m [hunk 3]
```diff
@@ -1,354 +0,0 @@
...
-    int currentValue;
-    int onesValue;
-    int fiveValue;
```
Comment (duckworthd) [inline sampleGame/GameScene.m]: Since it's very easy to derive `onesValue` and `fiveValue` from a single number, how about using one variable (e.g. `abbacusValue`) and defining methods that derive these two variables from it?

In a full implementation, it'd be nice to be able to have methods like `onesValue:(int)column` and `fivesValue:(int)column`.

File: sampleGame/GameScene.m [hunk 4]
```diff
@@ -1,354 +0,0 @@
...
-    //stores how many moves player has made
-    int moves;
-    
-    SKLabelNode *myLabel;
-    
-
-    NSMutableArray *positions; //stores original (all non-active) positions of beads
-    NSMutableArray *posBools;  //if beads are active (being counted) or not
-    
-    Abacus *_abacus;
```
Comment (duckworthd) [inline sampleGame/GameScene.m]: Don't forget to remove unused variables! Clean code is happy code :)

File: sampleGame/GameScene.m [hunk 5]
```diff
@@ -1,354 +0,0 @@
...
-    NSMutableArray *_onesArray;
-    
-    NSMutableArray *_moves; //stores value of column at each move
-}
-
-
--(void)didMoveToView:(SKView *)view {
```
Comment (duckworthd) [inline sampleGame/GameScene.m]: (Applies to all method definitions). Please add a declaration with the method name `GameScene.h` and comments describing their purpose, arguments, and return values. From the header alone, readers should be able to understand what the class is for and what each method does.

File: sampleGame/GameScene.m [hunk 6]
```diff
@@ -1,354 +0,0 @@
...
-    
-    myLabel.text = @"Hello, World!";
-    myLabel.fontSize = 35;
-    myLabel.position = CGPointMake(CGRectGetMidX(self.frame),
-                                   CGRectGetMidY(self.frame));
-    
-    [self addChild:myLabel];
-    
-    //intialize values, posBool array,, moves array
-    onesValue = 0;
```
Comment (duckworthd) [inline sampleGame/GameScene.m]: Initialize in `@implementation` where possible.

File: sampleGame/GameScene.m [hunk 7]
```diff
@@ -1,354 +0,0 @@
...
-    fiveValue = 0;
-    currentValue = 0;
```
Comment (duckworthd) [inline sampleGame/GameScene.m]: Initialize in `@implementation` where possible.

File: sampleGame/GameScene.m [hunk 8]
```diff
@@ -1,354 +0,0 @@
...
-    
-    posBools = [[NSMutableArray alloc] init];
-    posBools = [NSMutableArray arrayWithArray:@[@(0),@(0),@(0),@(0),@(0)]];
```
Comment (duckworthd) [inline sampleGame/GameScene.m]: Redundant with line immediately above. Please remove.

File: sampleGame/GameScene.m [hunk 9]
```diff
@@ -1,354 +0,0 @@
...
-    [self addChild:divider];
-    
-    //Setup beads
-    _columnBeads = [[NSMutableArray alloc] init];
-    positions = [[NSMutableArray alloc] init];
-    
-    
-    //Setup five bead
-    SKSpriteNode *fiveBead = [SKSpriteNode spriteNodeWithImageNamed:@"OrangeBead"];
-    fiveBead.position = CGPointMake(500,500);
```
Comment (duckworthd) [inline sampleGame/GameScene.m]: The viewable area of a scene is defined in `size` and `anchorPoint` (see [tutorial](https://developer.apple.com/library/ios/documentation/GraphicsAnimation/Conceptual/SpriteKit_PG/Nodes/Nodes.html)). It's dangerous to assume that `(500,500)` is viewable or centered.

File: sampleGame/GameScene.m [hunk 10]
```diff
@@ -1,354 +0,0 @@
...
-    fiveValue = 0;
-    currentValue = 0;
-    
-    posBools = [[NSMutableArray alloc] init];
-    posBools = [NSMutableArray arrayWithArray:@[@(0),@(0),@(0),@(0),@(0)]];
-    
-    _moves = [[NSMutableArray alloc] init];
-    [_moves addObject:[NSNumber numberWithInt:0]]; //so first move can compare to the starting value of 0
-    
-    //Setup Abacus
```
Comment (duckworthd) [inline sampleGame/GameScene.m]: Consider moving everything below here to an `initializeAbacusView` method. As the app develops, this should be moved to its own child-view independent of the `GameScene`.

File: sampleGame/GameScene.m [hunk 11]
```diff
@@ -1,354 +0,0 @@
...
-    
-    
-    //Setup five bead
-    SKSpriteNode *fiveBead = [SKSpriteNode spriteNodeWithImageNamed:@"OrangeBead"];
-    fiveBead.position = CGPointMake(500,500);
-    // NSLog(@"bead position x:%f y:%f",bead.position.x, bead.position.y);
-    [self addChild:fiveBead];
-    _columnBeads[0] = fiveBead;
-    fiveBead.name = [NSString stringWithFormat:@"%d", 5];
-    positions[0] = [NSNumber numberWithFloat:fiveBead.position.y];
```
Comment (duckworthd) [inline sampleGame/GameScene.m]: It appears that `positions[0]` is defined to be the position of the 5s bead. Having two variables implicitly be in sync is easy to mess up. Consider adding a method that derives `positions[0]` from `fiveBead` rather than storing it directly.

File: sampleGame/GameScene.m [hunk 12]
```diff
@@ -1,354 +0,0 @@
...
-        }
-    }
-    return beadNumber;
-}
-
-//Problem: trying to create better method with NSAssert & output parameters that first checks a bead IS being touched.
-//Example: http://www.raywenderlich.com/66877/how-to-make-a-game-like-candy-crush-part-1
-//Look at "convertPoint:toColumn:row:" method.
-//trying assert and output parameters
--(BOOL)whichBeadMovedwithTouch2:(UITouch *)touch{
```
Comment (duckworthd) [inline sampleGame/GameScene.m]: Please remove crufty code~

File: sampleGame/GameScene.m [hunk 13]
```diff
@@ -1,354 +0,0 @@
...
-    
-        
-        if (CGRectContainsPoint(boundingBox, touchLocation)) {
-            beadNumber = [NSNumber numberWithInt:(i)];
-            //to test:
-            //each bead's name is which bead they are i.e. 5 or the 1st/2nd/3rd/4th ones bead
-           // NSLog(@"touching bead: %@",testbead.name);
-        }
-    }
-    return beadNumber;
```
Comment (duckworthd) [inline sampleGame/GameScene.m]: Dangerous! Value of `beadNumber` is undefined if touch wasn't in any bounding box!

File: sampleGame/GameScene.m [hunk 14]
```diff
@@ -1,354 +0,0 @@
...
-        //  NSLog(@"bead position x:%f y:%f",bead.position.x, bead.position.y);
-        [self addChild:bead];
-        _columnBeads[i] = bead;
-        positions[i] = [NSNumber numberWithFloat:bead.position.y];
-    }
-}
-
-
-
--(NSNumber*)whichBeadMovedwithTouch:(UITouch *)touch{
```
Comment (duckworthd) [inline sampleGame/GameScene.m]: (Applies to all other methods) Add tests! Tests guarantee your code works! Though it feels slow, testing will make coding faster in the long run. Being able to trust code correctness >>> "moving fast", no matter what the techbois tell you.

File: sampleGame/GameScene.m [hunk 15]
```diff
@@ -1,354 +0,0 @@
...
-//Look at "convertPoint:toColumn:row:" method.
-//trying assert and output parameters
--(BOOL)whichBeadMovedwithTouch2:(UITouch *)touch{
-   // NSParameterAssert(beadNumber);
-    
-    CGPoint touchLocation = [touch locationInNode:self];
-    
-     SKNode *node = [self nodeAtPoint:touchLocation];
-    NSLog(@"node is: %@",node.name);
-    //returns bead.name is bead touched
```
Comment (duckworthd) [inline sampleGame/GameScene.m]: Remove commented-out code entirely. Git will save history for you to recover later if need be :)

File: sampleGame/GameScene.m [hunk 16]
```diff
@@ -1,354 +0,0 @@
...
-    //otherwise returns Null
-    
-    
-    if ([[self nodeAtPoint:touchLocation] name]){
-       // SKNode *node = [self nodeAtPoint:touchLocation];
-        NSLog(@"yay");
```
Comment (duckworthd) [inline sampleGame/GameScene.m]: Log statements should say what happened. This is fine for a quick check, but I bet you'll forget what it was for in a week :)

File: sampleGame/GameScene.m [hunk 17]
```diff
@@ -1,354 +0,0 @@
...
-    
-    CGPoint touchLocation = [touch locationInNode:self];
-    
-     SKNode *node = [self nodeAtPoint:touchLocation];
-    NSLog(@"node is: %@",node.name);
-    //returns bead.name is bead touched
-    //otherwise returns Null
-    
-    
-    if ([[self nodeAtPoint:touchLocation] name]){
```
Comment (duckworthd) [inline sampleGame/GameScene.m]: The entire if-else could be collapsed int `return [[self nodeAtPoint:touchLocation] name]`

File: sampleGame/GameScene.m (hunk 11 above)
Comment (duckworthd) [inline sampleGame/GameScene.m]: **EDIT** I now see that `_columnsBead` and `positions` together describe if a bead is active or not, and `posBools` stores that result. One array by itself can express this -- less is more! Please combine!

File: sampleGame/GameScene.m [hunk 18]
```diff
@@ -1,354 +0,0 @@
...
-        
-        //What is the first (lowest index) bead that will toggled?
-        int start = MIN(onesValue + 1, beadTappedIndex); //needs onesValue+1 because _columnBeads[0] is five-bead
-      
-        //How many beads will be toggled?
-        int length = MAX(onesValue, beadTappedIndex) - MIN(onesValue, beadTappedIndex - 1);
-       
-        //test with:
-        //for value = 4  to value = 2; length should be 2
-        //l = max(4,3) - min (4, 2) = 4-2 = 2
```
Comment (duckworthd) [inline sampleGame/GameScene.m]: Use this in a unit test!

File: sampleGame/GameScene.m [hunk 19]
```diff
@@ -1,354 +0,0 @@
...
-    //What is its current y-position?
-    CGFloat currentYPos = (CGFloat) bead.position.y;
-    
-    //What is its default (inactive) y-position?
-    float defaultYPos = [[positions objectAtIndex:beadIndex] floatValue];
-    
-    //if bead is in inactive (default) position > toggle to active position
-    CGFloat newY;
-    if (currentYPos == defaultYPos){
-        currentYPos += 50;
```
Comment (duckworthd) [inline sampleGame/GameScene.m]: Dangerous constant! Maybe you'd like `size / 20` or something frame-size-aware?

File: sampleGame/GameScene.m [hunk 20]
```diff
@@ -1,354 +0,0 @@
...
-    //Which bead was tapped?
-    int beadTappedIndex = beadTapped.intValue;
-    
-    
-    //If five-bead tapped
-    if (beadTappedIndex == 0){
-       // NSLog(@"five bead tapped");
-        //Subtract five bead, move up
-        if (fiveValue == 5){
-            [self moveBeadUp:@(0)];
```
Comment (duckworthd) [inline sampleGame/GameScene.m]: `toggleBeadPosition` is applicable here too.

File: sampleGame/GameScene.m [hunk 21]
```diff
@@ -1,354 +0,0 @@
...
-    CGFloat newY = (CGFloat)currentYPos;
-    SKAction *action = [SKAction moveToY:newY duration:0.2];
-    [touchedBead runAction:action];
-}
-
--(void)calculateColumnValue{
-  //  NSLog(@"staring to calculate value");
-    [self storePositionBits];
-    currentValue = [self calculateBeadValues].intValue;
-    NSLog(@"value is: %d with %d + %d", currentValue, fiveValue, onesValue);
```
Comment (duckworthd) [inline sampleGame/GameScene.m]: Good comment! Very useful. Much like.

File: sampleGame/GameScene.m [hunk 22]
```diff
@@ -1,354 +0,0 @@
...
-    else{
-     //   *beadNumber = NSNotFound;
-        return NO;
-    }
-    
-    //what else?
-    //bead should be own class and testing if typeOfClass is Bead
-}
-
--(void)storePositionBits{
```
Comment (duckworthd) [inline sampleGame/GameScene.m]: Rather than storing position bits, why not just return an array containing them? It'll ensure `posBools` is never out of sync with the Abacus state.

File: sampleGame/GameScene.m [hunk 23]
```diff
@@ -1,354 +0,0 @@
...
-    NSLog(@"value is: %d with %d + %d", currentValue, fiveValue, onesValue);
-    
-    //store value of move
-    [_moves addObject:[NSNumber numberWithInt:currentValue]];
-    //increment moves by 1
-    moves += 1;
-    
-    //log if discrepancy
-    if (_moves.count != moves + 1){
-    NSLog(@"Warning: moves are not equal, moveArray is %d, movesCounter is %d", _moves.count, moves);
```
Comment (duckworthd) [inline sampleGame/GameScene.m]: OOoooooooo KatKat! Consistent tabbing plz.

File: sampleGame/GameScene.m [hunk 24]
```diff
@@ -1,354 +0,0 @@
...
-    }
-    NSLog(@"moveArray: %@", _moves);
-    
-    //What is the change in value comparing previous to current move/board?
-    int prevValIndex = _moves.count - 2;
-    int prevVal = [_moves[prevValIndex] intValue];
```
Comment (duckworthd) [inline sampleGame/GameScene.m]: Combine two lines. Also, what if `_moves.count == 0` or `1`?

File: sampleGame/GameScene.m [hunk 25]
```diff
@@ -1,354 +0,0 @@
...
-            //   [self calculateColumnValue]; //when called by itself, always reports last (not current) move. it needs delay, see fix in next line.
-            [self performSelector:@selector(calculateColumnValue) withObject:self afterDelay:0.4]; //needs delay because bead moving action takes 0.2 (should couple moveBeads action and this)
-        }
-        
-    }
-}
-
-
-
--(void)update:(CFTimeInterval)currentTime {
```
Comment (duckworthd) [inline sampleGame/GameScene.m]: Since this isn't performing anything, it can be removed?

File: sampleGame/GameScene.h [hunk 26]
```diff
@@ -1,12 +0,0 @@
-//
-//  GameScene.h
-//  sampleGame
-//
-
-//  Copyright (c) 2015 Katrina DeVaney. All rights reserved.
-//
```
Comment (duckworthd) [inline sampleGame/GameScene.h]: Please add comment describing what is in the scene and when it will be displayed.

**Comment by duckworthd** [comment]
It's ALIVE MUWAHAHAHAH :D :D :D 

Overall, good start! I left lots of little nitpicks about style and spacing and naming and combining variables, but it's headed in a very good direction.

Looking forward, I think a very good next step is,
1. Add unit tests. It will give you confidence in code correctness. Pay particular attention to edge cases that might make something crash or do something undefined.
2. Reduce the number of instance variables. Fewer variables means fewer mistakes!
3. Move as much Abacus-level view control logic into a child-view. In the full Appacus app, you'll want many abacus columns, after all.

Till next time! duckibot, out *bzzt*.

**Comment by kattak** [comment]
thank you duckibot for your generous beebawping.

i will learn how to make tests <3

On Mon, May 4, 2015 at 12:04 AM, Daniel Duckworth <notifications@github.com>
wrote:

> It's ALIVE MUWAHAHAHAH :D :D :D
>
> Overall, good start! I left lots of little nitpicks about style and
> spacing and naming and combining variables, but it's headed in a very good
> direction.
>
> Looking forward, I think a very good next step is,
> 1. Add unit tests. It will give you confidence in code correctness. Pay
> particular attention to edge cases that might make something crash or do
> something undefined.
> 2. Reduce the number of instance variables. Fewer variables means fewer
> mistakes!
> 3. Move as much Abacus-level view control logic into a child-view. In the
> full Appacus app, you'll want many abacus columns, after all.
>
> Till next time! duckibot, out *bzzt*.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kattak/sampleGame/pull/1#issuecomment-98603888>.
>



-- 
Katrina DeVaney


University of California, Berkeley l May 2014

Cognitive Science <http://en.wikipedia.org/wiki/Cognitive_science>
B.A. | Education Minor


E: katrina.devaney@gmail.com | C: 773.679.9667
